File: lld/week-02-03-patterns/singleton/go/singleton_test.go
package singleton

import (
	"sync"
	"testing"
)

func TestGetInstanceReturnsSameInstance(t *testing.T) {
	resetForTest()
	a := GetInstance()
	b := GetInstance()
	if a != b {
		t.Fatalf("expected same pointer, got different instances")
	}
	if a.ID() != b.ID() {
		t.Fatalf("expected same id, got %d and %d", a.ID(), b.ID())
	}
}

func TestSetAndGet(t *testing.T) {
	resetForTest()
	cfg := GetInstance()
	cfg.Set("env", "production")

	v, ok := cfg.Get("env")
	if !ok || v != "production" {
		t.Fatalf("expected env=production, got %q ok=%v", v, ok)
	}

	if _, ok := cfg.Get("missing"); ok {
		t.Fatalf("expected missing key to be absent")
	}
}

// TestConcurrentFirstAccess spins up many goroutines racing to call
// GetInstance for the first time; sync.Once must guarantee exactly one
// AppConfig gets constructed, so every goroutine must observe the same id.
func TestConcurrentFirstAccess(t *testing.T) {
	resetForTest()

	const goroutines = 200
	var wg sync.WaitGroup
	ids := make([]int, goroutines)

	for i := 0; i < goroutines; i++ {
		wg.Add(1)
		go func(idx int) {
			defer wg.Done()
			ids[idx] = GetInstance().ID()
		}(i)
	}
	wg.Wait()

	first := ids[0]
	for i, id := range ids {
		if id != first {
			t.Fatalf("goroutine %d observed id %d, expected %d (multiple instances constructed)", i, id, first)
		}
	}
}

func TestDemo(t *testing.T) {
	resetForTest()
	cfg := GetInstance()
	cfg.Set("region", "us-east-1")
	v, _ := cfg.Get("region")
	if v != "us-east-1" {
		t.Fatalf("demo failed, got %q", v)
	}
}

File: lld/week-02-03-patterns/singleton/java/src/SingletonTest.java
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicIntegerArray;

/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out SingletonTest` directly.
 */
public class SingletonTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testSameInstanceReturned();
        testSetAndGet();
        testConcurrentFirstAccess();
        System.out.println("All SingletonTest cases passed.");
    }

    private static void testSameInstanceReturned() {
        AppConfig a = AppConfig.getInstance();
        AppConfig b = AppConfig.getInstance();
        assertEquals(true, a == b, "getInstance must return the same reference");
        assertEquals(a.getId(), b.getId(), "instance ids must match");
    }

    private static void testSetAndGet() {
        AppConfig config = AppConfig.getInstance();
        config.set("region", "us-east-1");
        assertEquals("us-east-1", config.get("region"), "set/get roundtrip");
    }

    private static void testConcurrentFirstAccess() {
        final int threads = 50;
        AtomicIntegerArray ids = new AtomicIntegerArray(threads);
        CountDownLatch latch = new CountDownLatch(threads);

        for (int i = 0; i < threads; i++) {
            final int idx = i;
            new Thread(() -> {
                ids.set(idx, AppConfig.getInstance().getId());
                latch.countDown();
            }).start();
        }
        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        int first = ids.get(0);
        for (int i = 0; i < threads; i++) {
            assertEquals(first, ids.get(i), "every thread must observe the same singleton id");
        }
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
